### PR TITLE
Use a single network connection for reading and writing

### PR DESCRIPTION
**Pull request by jack0**

### CODE REVIEW

File: network/network.go [hunk 1]
```diff
@@ -235,8 +235,9 @@ func (n *Network) Listen() {
 	}
 }
 
-// Client either creates or returns a cached peer client given its host address.
-func (n *Network) Client(address string) (*PeerClient, error) {
+// initClient either returns a cached peer client or creates a new one given a net.Conn
+// or dials the client if no net.Conn is provided.
+func (n *Network) initClient(address string, conn net.Conn) (*PeerClient, error) {
```
Comment (doug-perlin) [inline network/network.go]: might rename this `getOrSetPeerClient()`, init seems like it can only be done once

**Comment by doug-perlin** [comment]
Might have to change this to `no` to fix the codecov/patch:
https://github.com/perlin-network/noise/blob/master/.codecov.yml#L25

**Comment by jack0** [comment]
can't they still do that in the old code? all they need to do is change `Sender` in the PrepareMessage?

**Comment by losfair** [comment]
In the old code, the remote identity (at least the remote address) is verified by always actively connecting for the write channel.

File: network/network.go [hunk 2]
```diff
@@ -246,6 +248,23 @@ func (n *Network) Client(address string) (*PeerClient, error) {
...
+		if err != nil {
+			return nil, err
+		}
+
+		remoteAddrInfo, err := ParseAddress(fmt.Sprintf("%s://%s", conn.RemoteAddr().Network(), conn.RemoteAddr().String()))
+		if err != nil {
+			return nil, err
+		}
+
+		if addrInfo.Host != remoteAddrInfo.Host {
```
Comment (losfair) [inline network/network.go]: What about NAT if we only check the host?

**Pull request merged**